api/restHandler/GitHostRestHandler.go

```diff
@@ -0,0 +1,168 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package restHandler
+
+import (
+	"encoding/json"
+	"github.com/devtron-labs/devtron/pkg/pipeline"
+	"github.com/devtron-labs/devtron/pkg/user"
+	"github.com/devtron-labs/devtron/util/rbac"
+	"github.com/gorilla/mux"
+	"go.uber.org/zap"
+	"gopkg.in/go-playground/validator.v9"
+	"net/http"
+	"strconv"
+	"strings"
+)
+
+type GitHostRestHandler interface {
+	GetGitHosts(w http.ResponseWriter, r *http.Request)
+	GetGitHostById(w http.ResponseWriter, r *http.Request)
+	CreateGitHost(w http.ResponseWriter, r *http.Request)
+}
+
+type GitHostRestHandlerImpl struct {
+	logger               *zap.SugaredLogger
+	gitHostConfig    	 pipeline.GitHostConfig
+	userAuthService      user.UserService
+	validator            *validator.Validate
+	enforcer             rbac.Enforcer
+}
+
+func NewGitHostRestHandlerImpl(logger *zap.SugaredLogger,
+	gitHostConfig pipeline.GitHostConfig, userAuthService user.UserService,
+	validator *validator.Validate, enforcer rbac.Enforcer) *GitHostRestHandlerImpl {
+	return &GitHostRestHandlerImpl{
+		logger:               logger,
+		gitHostConfig:    gitHostConfig,
+		userAuthService: userAuthService,
+		validator: validator,
+		enforcer: enforcer,
+	}
+}
+
+func (impl GitHostRestHandlerImpl) GetGitHosts(w http.ResponseWriter, r *http.Request) {
+
+	// check if user is logged in or not
+	userId, err := impl.userAuthService.GetLoggedInUser(r)
+	if userId == 0 || err != nil {
+		writeJsonResp(w, err, "Unauthorized User", http.StatusUnauthorized)
+		return
+	}
+
+	res, err := impl.gitHostConfig.GetAll()
+	if err != nil {
+		impl.logger.Errorw("service err, GetGitHosts", "err", err)
+		writeJsonResp(w, err, nil, http.StatusInternalServerError)
+		return
+	}
+
+	// RBAC enforcer applying
+	token := r.Header.Get("token")
+	var result []pipeline.GitHostRequest
+	for _, item := range res {
+		if ok := impl.enforcer.Enforce(token, rbac.ResourceGit, rbac.ActionGet, strings.ToLower(item.Name)); ok {
+			result = append(result, item)
+		}
+	}
+	//RBAC enforcer Ends
+
+	writeJsonResp(w, err, res, http.StatusOK)
+}
+
+func (impl GitHostRestHandlerImpl) GetGitHostById(w http.ResponseWriter, r *http.Request) {
+
+	// check if user is logged in or not
+	userId, err := impl.userAuthService.GetLoggedInUser(r)
+	if userId == 0 || err != nil {
+		writeJsonResp(w, err, "Unauthorized User", http.StatusUnauthorized)
+		return
+	}
+
+	params := mux.Vars(r)
+	id, err :=  strconv.Atoi(params["id"])
+
+	if err != nil {
+		impl.logger.Errorw("service err in parsing Id , GetGitHostById", "err", err)
+		writeJsonResp(w, err, nil, http.StatusInternalServerError)
+		return
+	}
+
+	res, err := impl.gitHostConfig.GetById(id)
+
+	if err != nil {
+		impl.logger.Errorw("service err, GetGitHostById", "err", err)
+		writeJsonResp(w, err, nil, http.StatusInternalServerError)
```

Comment:
will this error be understandable for user?

Reply:
Copied this logic from other rest handlers. I know this is not a proper error message. but we should fix for all the APIs following a common mechanism

Reply:
agree lets start  by fixing here